<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#PHM2017-Dataset-Basic-Exploration" data-toc-modified-id="PHM2017-Dataset-Basic-Exploration-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>PHM2017 Dataset Basic Exploration</a></span></li><li><span><a href="#clean-the-dataset" data-toc-modified-id="clean-the-dataset-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>clean the dataset</a></span></li><li><span><a href="#load-word-embeddings-from-the-Glove" data-toc-modified-id="load-word-embeddings-from-the-Glove-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>load word embeddings from the Glove</a></span></li><li><span><a href="#prepare-dataset" data-toc-modified-id="prepare-dataset-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>prepare dataset</a></span><ul class="toc-item"><li><span><a href="#change-df_to_evaluate-to-eval-different-dataset" data-toc-modified-id="change-df_to_evaluate-to-eval-different-dataset-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>change df_to_evaluate to eval different dataset</a></span></li><li><span><a href="#functions-to-be-used" data-toc-modified-id="functions-to-be-used-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>functions to be used</a></span></li></ul></li><li><span><a href="#baseline-experiment" data-toc-modified-id="baseline-experiment-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>baseline experiment</a></span><ul class="toc-item"><li><span><a href="#MLP-keras-version" data-toc-modified-id="MLP-keras-version-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>MLP-keras version</a></span></li><li><span><a href="#bi-lstm" data-toc-modified-id="bi-lstm-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>bi-lstm</a></span></li><li><span><a href="#GRU+glove" data-toc-modified-id="GRU+glove-5.3"><span class="toc-item-num">5.3&nbsp;&nbsp;</span>GRU+glove</a></span></li><li><span><a href="#LSTM+glove" data-toc-modified-id="LSTM+glove-5.4"><span class="toc-item-num">5.4&nbsp;&nbsp;</span>LSTM+glove</a></span></li><li><span><a href="#CNN+glove" data-toc-modified-id="CNN+glove-5.5"><span class="toc-item-num">5.5&nbsp;&nbsp;</span>CNN+glove</a></span></li><li><span><a href="#char-embedding" data-toc-modified-id="char-embedding-5.6"><span class="toc-item-num">5.6&nbsp;&nbsp;</span>char-embedding</a></span><ul class="toc-item"><li><span><a href="#cnn-char-model-building" data-toc-modified-id="cnn-char-model-building-5.6.1"><span class="toc-item-num">5.6.1&nbsp;&nbsp;</span>cnn-char model building</a></span></li><li><span><a href="#lstm-(integrate-cnn-char)-model" data-toc-modified-id="lstm-(integrate-cnn-char)-model-5.6.2"><span class="toc-item-num">5.6.2&nbsp;&nbsp;</span>lstm (integrate cnn-char) model</a></span></li><li><span><a href="#lstm-(integrate-rnn-char)-model" data-toc-modified-id="lstm-(integrate-rnn-char)-model-5.6.3"><span class="toc-item-num">5.6.3&nbsp;&nbsp;</span>lstm (integrate rnn-char) model</a></span></li></ul></li></ul></li><li><span><a href="#result-table" data-toc-modified-id="result-table-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>result table</a></span><ul class="toc-item"><li><span><a href="#old-result-of-binary-version" data-toc-modified-id="old-result-of-binary-version-6.1"><span class="toc-item-num">6.1&nbsp;&nbsp;</span>old result of binary version</a></span><ul class="toc-item"><li><span><a href="#f1-score-of-positive-class" data-toc-modified-id="f1-score-of-positive-class-6.1.1"><span class="toc-item-num">6.1.1&nbsp;&nbsp;</span>f1 score of positive class</a></span></li><li><span><a href="#macro-average-of-f1-score" data-toc-modified-id="macro-average-of-f1-score-6.1.2"><span class="toc-item-num">6.1.2&nbsp;&nbsp;</span>macro average of f1 score</a></span></li></ul></li><li><span><a href="#new-result-of-4-classes" data-toc-modified-id="new-result-of-4-classes-6.2"><span class="toc-item-num">6.2&nbsp;&nbsp;</span>new result of 4 classes</a></span></li></ul></li><li><span><a href="#some-notes-in-training:" data-toc-modified-id="some-notes-in-training:-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>some notes in training:</a></span></li><li><span><a href="#conclusion" data-toc-modified-id="conclusion-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>conclusion</a></span></li></ul></div>

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
pd.options.display.max_columns = None

In [2]:
# Keras
from keras.preprocessing.text import Tokenizer
from keras.preprocessing.sequence import pad_sequences
from keras.models import Sequential
from keras.layers import Dense, Flatten, LSTM, Conv1D, MaxPooling1D, Dropout, Activation,GRU
from keras.layers.embeddings import Embedding

# Others
import nltk
import string
import numpy as np
import pandas as pd
from nltk.corpus import stopwords
from sklearn.manifold import TSNE
from keras.utils import np_utils
from sklearn.preprocessing import LabelEncoder

D:\softwares\Anaconda3\lib\site-packages\h5py\__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
Using TensorFlow backend.


# PHM2017 Dataset Basic Exploration

for more detailed exploration, please refer to PHM2017.nb

In [3]:
df=pd.read_csv('../clean_data/PHM2017/PHM2017.csv')
print(df.shape)
df.head()

(5288, 4)


id                                              tweet  \
0  811644183937159168  Had a sleep as I went for 24hrs with no sleep....   
1  811644622904721408  Owning a cat can reduce the risk of a stroke a...   
2  811644687899652100         phew. U nearly gave me a heart attack then   
3  811644691619991552  @emmabennettx a cheeky reply and you'll give t...   
4  811644742148702208  A linesman's whistle induced heart attack and ...   

        symptom  label  
0  heart attack      1  
1  heart attack      1  
2  heart attack      0  
3  heart attack      0  
4  heart attack      0

# clean the dataset

In [4]:
import string
from nltk.corpus import stopwords
import re
from nltk.stem import SnowballStemmer

def clean_text(text):
    
    ## Remove puncuation
    text = text.translate(string.punctuation)
    
    ## Convert words to lower case and split them
    text = text.lower().split()
    
    ## Remove stop words
    stops = set(stopwords.words("english"))
    text = [w for w in text if not w in stops and len(w) >= 3]
    
    text = " ".join(text)
    ## Clean the text
    text = re.sub(r"[^A-Za-z0-9^,!.\/'+-=]", " ", text)
    text = re.sub(r"what's", "what is ", text)
    text = re.sub(r"\'s", " ", text)
    text = re.sub(r"\'ve", " have ", text)
    text = re.sub(r"n't", " not ", text)
    text = re.sub(r"i'm", "i am ", text)
    text = re.sub(r"\'re", " are ", text)
    text = re.sub(r"\'d", " would ", text)
    text = re.sub(r"\'ll", " will ", text)
    text = re.sub(r",", " ", text)
    text = re.sub(r"\.", " ", text)
    text = re.sub(r"!", " ! ", text)
    text = re.sub(r"\/", " ", text)
    text = re.sub(r"\^", " ^ ", text)
    text = re.sub(r"\+", " + ", text)
    text = re.sub(r"\-", " - ", text)
    text = re.sub(r"\=", " = ", text)
    text = re.sub(r"'", " ", text)
    text = re.sub(r"(\d+)(k)", r"\g<1>000", text)
    text = re.sub(r":", " : ", text)
    text = re.sub(r" e g ", " eg ", text)
    text = re.sub(r" b g ", " bg ", text)
    text = re.sub(r" u s ", " american ", text)
    text = re.sub(r"\0s", "0", text)
    text = re.sub(r" 9 11 ", "911", text)
    text = re.sub(r"e - mail", "email", text)
    text = re.sub(r"j k", "jk", text)
    text = re.sub(r"\s{2,}", " ", text)
    ## Stemming
    text = text.split()
    stemmer = SnowballStemmer('english')
    stemmed_words = [stemmer.stem(word) for word in text]
    text = " ".join(stemmed_words)
    return text
# apply the above function to df['text']
df['tweet'] = df['tweet'].map(lambda x: clean_text(x))
df.head()

id                                              tweet  \
0  811644183937159168  sleep went 24hrs sleep esophogus sore yeah giv...   
1  811644622904721408       own cat reduc risk stroke heart attack third   
2  811644687899652100                        phew near gave heart attack   
3  811644691619991552  emmabennettx cheeki repli give poor bastard so...   
4  811644742148702208  linesman whistl induc heart attack nice video ...   

        symptom  label  
0  heart attack      1  
1  heart attack      1  
2  heart attack      0  
3  heart attack      0  
4  heart attack      0

In [5]:
m=0
for s in df['tweet'].tolist():
    if(len(s.split())>m):
#         print(s.split())
        m=len(s.split())
print('max sentence length:',m)

### Create sequence
vocabulary_size = 11401
tokenizer = Tokenizer(num_words= vocabulary_size)
tokenizer.fit_on_texts(df['tweet'])
sequences = tokenizer.texts_to_sequences(df['tweet'])
data = pad_sequences(sequences, maxlen=100)

max sentence length: 87


In [6]:
data.shape

(5288, 100)

# load word embeddings from the Glove

In [7]:
embeddings_index = dict()
f = open('D:/Datasets/glove.6B.300d.txt','r',encoding='utf-8')
for line in f:
    values = line.split()
    word = values[0]
    coefs = np.asarray(values[1:], dtype='float32')
    embeddings_index[word] = coefs
f.close()

embedding_matrix = np.zeros((vocabulary_size, 300))
for word, index in tokenizer.word_index.items():
    if index > vocabulary_size - 1:
        break
    else:
        embedding_vector = embeddings_index.get(word)
        if embedding_vector is not None:
            embedding_matrix[index] = embedding_vector

# prepare dataset

In [8]:
df_alzheimer=df[df['symptom']=='alzheimer'][['tweet','label']]
df_cancer=df[df['symptom']=='cancer'][['tweet','label']]
df_depression=df[df['symptom']=='depression'][['tweet','label']]
df_heart_attack=df[df['symptom']=='heart attack'][['tweet','label']]
df_parkinson=df[df['symptom']=='parkinson'][['tweet','label']]
df_stroke=df[df['symptom']=='stroke'][['tweet','label']]

## change df_to_evaluate to eval different dataset

In [30]:
# change df_to_evaluate to evaluate different types

df_to_evaluate=df_stroke

stroke_train=data[df_to_evaluate.index]
stroke_label=df_to_evaluate.label.tolist()

## functions to be used

In [10]:

def recall_m(y_true, y_pred):
        true_positives = K.sum(K.round(K.clip(y_true * y_pred, 0, 1)))
        possible_positives = K.sum(K.round(K.clip(y_true, 0, 1)))
        recall = true_positives / (possible_positives + K.epsilon())
        return recall

def precision_m(y_true, y_pred):
        true_positives = K.sum(K.round(K.clip(y_true * y_pred, 0, 1)))
        predicted_positives = K.sum(K.round(K.clip(y_pred, 0, 1)))
        precision = true_positives / (predicted_positives + K.epsilon())
        return precision

def f1_m(y_true, y_pred):
    precision = precision_m(y_true, y_pred)
    recall = recall_m(y_true, y_pred)
    return 2*((precision*recall)/(precision+recall+K.epsilon()))


# baseline experiment

In [11]:
## MLP+glove

In [359]:
dataset=[]
for t in df_to_evaluate.tweet.tolist():
    dataset.append(t.split())
    
EMBEDDING_DIM=300
data_emb=np.zeros((len(dataset),EMBEDDING_DIM))
for i in range(len(dataset)):
    each_emb=np.zeros((len(dataset[i]),EMBEDDING_DIM))
    for j,w in enumerate(dataset[i]):
        if(w in embeddings_index):
            each_emb[j]=embeddings_index[w]
#             break
    data_emb[i]=np.mean(each_emb, axis=0)        
        
#     nonzero_inds=np.nonzero(dataset[i].flatten())
#     review_emb=embedding_matrix[nonzero_inds]
#     data_emb[i]=np.mean(review_emb, axis=0)
data_emb.shape

(890, 300)

In [360]:
# from sklearn.neural_network import MLPClassifier
# from sklearn.model_selection import cross_val_score
# from sklearn.model_selection import cross_validate

# hidden_neurons=256
    
# classifier = MLPClassifier(hidden_layer_sizes=(hidden_neurons,hidden_neurons), max_iter=1000, alpha=0.001,
#                      solver='adam', verbose=0,  random_state=0)

# print("start training..")
# # classifier.fit(X_train_emb,y_train)
# scoring = ['precision_macro', 'recall_macro','f1_macro']
# scores = cross_validate(classifier, data_emb, df_to_evaluate.label, 
# #                         print evaluation metric for each class
# #                          scoring=make_scorer(classification_report_with_accuracy_score),
#                          scoring=scoring,
#                          cv=5)
# print(scores)
# print('ave_test_f1_macro:',scores['test_f1_macro'].mean())

In [17]:
# old result
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
from sklearn.model_selection import train_test_split

X_train, X_val_test, y_train, y_val_test = train_test_split(data_emb, stroke_label,
                                                    stratify=stroke_label, 
                                                    test_size=0.4,
                                                    random_state=0
                                                   )
X_val,X_test,y_val,y_test = train_test_split(X_val_test, y_val_test,
                                                    stratify=y_val_test, 
                                                    test_size=0.5,
                                                    random_state=0
                                                   )
y_train_cat= np_utils.to_categorical(y_train)
y_test_cat= np_utils.to_categorical(y_test)
y_val_cat= np_utils.to_categorical(y_val)

print(X_train.shape,X_val_test.shape,X_val.shape,X_test.shape)
# classifier.fit(X_train,y_train)

# y_pred = classifier.predict(X_test)
# # y_pred = y_prob.argmax(axis=-1)

# y_test


NameError: name 'data_emb' is not defined

## MLP-keras version

In [16]:
from keras.models import Sequential
from keras.layers import Dense, Activation
from keras import backend as K

model_mlp = Sequential([
    Dense(128, input_shape=(300,)),
    Activation('relu'),
    Dense(4),
    Activation('softmax'),
])

model_mlp.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['acc',f1_m,precision_m, recall_m])
# set class_weight for imbalanced classes
# class_weight={0:1,1:1,2:10,3:10}
class_weight={0:1,1:1,2:10,3:10}


model_mlp.fit(X_train, y_train_cat, validation_data=(X_val, y_val_cat), epochs = 50,class_weight=class_weight)

ValueError: Error when checking input: expected dense_3_input to have shape (300,) but got array with shape (100,)

In [363]:

y_prob = model_mlp.predict(X_test)
y_pred = y_prob.argmax(axis=-1)

confusion_matrix(y_test, y_pred)
print(classification_report(y_test, y_pred))
# can use macro avg as evaluation metric

              precision    recall  f1-score   support

           0       0.81      0.85      0.83        98
           1       0.77      0.70      0.73        53
           2       0.64      0.70      0.67        23
           3       0.00      0.00      0.00         4

   micro avg       0.76      0.76      0.76       178
   macro avg       0.55      0.56      0.56       178
weighted avg       0.76      0.76      0.76       178



## bi-lstm

In [31]:

from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
from sklearn.model_selection import train_test_split
from keras import backend as K

X_train, X_val_test, y_train, y_val_test = train_test_split(stroke_train, stroke_label,
                                                    stratify=stroke_label, 
                                                    test_size=0.4,
                                                    random_state=0
                                                   )

X_val,X_test,y_val,y_test = train_test_split(X_val_test, y_val_test,
                                                    stratify=y_val_test, 
                                                    test_size=0.5,
                                                    random_state=0
                                                   )
y_train_cat= np_utils.to_categorical(y_train)
y_test_cat= np_utils.to_categorical(y_test)
y_val_cat= np_utils.to_categorical(y_val)


In [32]:
X_train.shape,X_test.shape,y_train_cat.shape,y_test_cat.shape,

((534, 100), (178, 100), (534, 4), (178, 4))

In [33]:
from keras.layers import Bidirectional

model_glove = Sequential()
model_glove.add(Embedding(vocabulary_size, 300, input_length=100, weights=[embedding_matrix], trainable=False))
model_glove.add(Dropout(0.5))
# model_glove.add(Conv1D(64, 5, activation='relu'))
# model_glove.add(MaxPooling1D(pool_size=4))
model_glove.add(Bidirectional(LSTM(100)))
model_glove.add(Dropout(0.2))
model_glove.add(Dense(4, activation='softmax'))
# model_glove.add(Dense(4, activation='sigmoid'))
model_glove.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['acc',f1_m,precision_m, recall_m])
## Fit train data

# set class_weight for imbalanced classes
class_weight={0:1,1:1,2:10,3:10}


model_glove.fit(X_train, y_train_cat, validation_data=(X_val, y_val_cat), epochs = 100,class_weight=class_weight)

Train on 534 samples, validate on 178 samples
Epoch 1/100
534/534 [==============================] - 4s 8ms/step - loss: 2.9268 - acc: 0.1292 - f1_m: 0.0480 - precision_m: 0.1432 - recall_m: 0.0337 - val_loss: 1.6399 - val_acc: 0.1348 - val_f1_m: 0.1168 - val_precision_m: 0.1392 - val_recall_m: 0.1011
Epoch 2/100
534/534 [==============================] - 2s 3ms/step - loss: 2.5739 - acc: 0.1517 - f1_m: 0.1120 - precision_m: 0.2011 - recall_m: 0.0805 - val_loss: 1.4324 - val_acc: 0.1910 - val_f1_m: 0.1169 - val_precision_m: 0.1706 - val_recall_m: 0.0899
Epoch 3/100
534/534 [==============================] - 2s 4ms/step - loss: 2.4103 - acc: 0.2060 - f1_m: 0.1448 - precision_m: 0.2280 - recall_m: 0.1067 - val_loss: 1.3827 - val_acc: 0.2528 - val_f1_m: 0.1718 - val_precision_m: 0.2404 - val_recall_m: 0.1348
Epoch 4/100
534/534 [==============================] - 2s 4ms/step - loss: 2.1827 - acc: 0.3596 - f1_m: 0.2311 - precision_m: 0.3743 - recall_m: 0.1704 - val_loss: 1.1368 - val_acc: 0

Epoch 33/100
534/534 [==============================] - 2s 4ms/step - loss: 0.3129 - acc: 0.9157 - f1_m: 0.9173 - precision_m: 0.9290 - recall_m: 0.9064 - val_loss: 0.9189 - val_acc: 0.7640 - val_f1_m: 0.7647 - val_precision_m: 0.7711 - val_recall_m: 0.7584
Epoch 34/100
534/534 [==============================] - 2s 4ms/step - loss: 0.2743 - acc: 0.9307 - f1_m: 0.9327 - precision_m: 0.9445 - recall_m: 0.9213 - val_loss: 0.9455 - val_acc: 0.7978 - val_f1_m: 0.7996 - val_precision_m: 0.8075 - val_recall_m: 0.7921
Epoch 35/100
534/534 [==============================] - 2s 4ms/step - loss: 0.3271 - acc: 0.9326 - f1_m: 0.9322 - precision_m: 0.9459 - recall_m: 0.9195 - val_loss: 0.9927 - val_acc: 0.7416 - val_f1_m: 0.7507 - val_precision_m: 0.7604 - val_recall_m: 0.7416
Epoch 36/100
534/534 [==============================] - 2s 4ms/step - loss: 0.2419 - acc: 0.9457 - f1_m: 0.9419 - precision_m: 0.9537 - recall_m: 0.9307 - val_loss: 1.0158 - val_acc: 0.7584 - val_f1_m: 0.7541 - val_precision_m

Epoch 65/100
534/534 [==============================] - 2s 4ms/step - loss: 0.1017 - acc: 0.9794 - f1_m: 0.9793 - precision_m: 0.9831 - recall_m: 0.9757 - val_loss: 1.1248 - val_acc: 0.7640 - val_f1_m: 0.7627 - val_precision_m: 0.7671 - val_recall_m: 0.7584
Epoch 66/100
534/534 [==============================] - 2s 4ms/step - loss: 0.1377 - acc: 0.9700 - f1_m: 0.9709 - precision_m: 0.9737 - recall_m: 0.9682 - val_loss: 1.2480 - val_acc: 0.7303 - val_f1_m: 0.7383 - val_precision_m: 0.7467 - val_recall_m: 0.7303
Epoch 67/100
534/534 [==============================] - 2s 4ms/step - loss: 0.1055 - acc: 0.9757 - f1_m: 0.9756 - precision_m: 0.9775 - recall_m: 0.9738 - val_loss: 1.1454 - val_acc: 0.7640 - val_f1_m: 0.7652 - val_precision_m: 0.7722 - val_recall_m: 0.7584
Epoch 68/100
534/534 [==============================] - 2s 4ms/step - loss: 0.0881 - acc: 0.9738 - f1_m: 0.9727 - precision_m: 0.9754 - recall_m: 0.9700 - val_loss: 1.1344 - val_acc: 0.7472 - val_f1_m: 0.7516 - val_precision_m

Epoch 97/100
534/534 [==============================] - 2s 4ms/step - loss: 0.0293 - acc: 1.0000 - f1_m: 1.0000 - precision_m: 1.0000 - recall_m: 1.0000 - val_loss: 1.3389 - val_acc: 0.7697 - val_f1_m: 0.7717 - val_precision_m: 0.7738 - val_recall_m: 0.7697
Epoch 98/100
534/534 [==============================] - 2s 4ms/step - loss: 0.0229 - acc: 0.9963 - f1_m: 0.9972 - precision_m: 0.9981 - recall_m: 0.9963 - val_loss: 1.3507 - val_acc: 0.7753 - val_f1_m: 0.7717 - val_precision_m: 0.7738 - val_recall_m: 0.7697
Epoch 99/100
534/534 [==============================] - 3s 5ms/step - loss: 0.0307 - acc: 0.9963 - f1_m: 0.9971 - precision_m: 0.9981 - recall_m: 0.9963 - val_loss: 1.3456 - val_acc: 0.7753 - val_f1_m: 0.7753 - val_precision_m: 0.7753 - val_recall_m: 0.7753
Epoch 100/100
534/534 [==============================] - 2s 4ms/step - loss: 0.0294 - acc: 0.9925 - f1_m: 0.9934 - precision_m: 0.9943 - recall_m: 0.9925 - val_loss: 1.3560 - val_acc: 0.7753 - val_f1_m: 0.7774 - val_precision_

In [34]:
## from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix

y_prob = model_glove.predict(X_test)
y_pred = y_prob.argmax(axis=-1)

confusion_matrix(y_test, y_pred)
print(classification_report(y_test, y_pred))
print(confusion_matrix(y_test, y_pred))
# can use macro avg as evaluation metric

              precision    recall  f1-score   support

           0       0.79      0.92      0.85        98
           1       0.90      0.68      0.77        53
           2       0.62      0.65      0.64        23
           3       0.00      0.00      0.00         4

   micro avg       0.79      0.79      0.79       178
   macro avg       0.58      0.56      0.57       178
weighted avg       0.78      0.79      0.78       178

[[90  3  5  0]
 [14 36  3  0]
 [ 7  1 15  0]
 [ 3  0  1  0]]


D:\softwares\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1143: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)
D:\softwares\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1143: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)
D:\softwares\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1143: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)


## GRU+glove

In [57]:

from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
from sklearn.model_selection import train_test_split
from keras import backend as K

X_train, X_val_test, y_train, y_val_test = train_test_split(stroke_train, stroke_label,
                                                    stratify=stroke_label, 
                                                    test_size=0.4,
                                                    random_state=0
                                                   )

X_val,X_test,y_val,y_test = train_test_split(X_val_test, y_val_test,
                                                    stratify=y_val_test, 
                                                    test_size=0.5,
                                                    random_state=0
                                                   )
y_train_cat= np_utils.to_categorical(y_train)
y_test_cat= np_utils.to_categorical(y_test)
y_val_cat= np_utils.to_categorical(y_val)


In [58]:
X_train.shape,X_test.shape,y_train_cat.shape,y_test_cat.shape,

((534, 100), (178, 100), (534, 4), (178, 4))

In [59]:

model_glove = Sequential()
model_glove.add(Embedding(vocabulary_size, 300, input_length=100, weights=[embedding_matrix], trainable=False))
model_glove.add(Dropout(0.2))
# model_glove.add(Conv1D(64, 5, activation='relu'))
# model_glove.add(MaxPooling1D(pool_size=4))
model_glove.add(GRU(100))
model_glove.add(Dropout(0.4))
model_glove.add(Dense(4, activation='softmax'))
# model_glove.add(Dense(4, activation='sigmoid'))
model_glove.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['acc',f1_m,precision_m, recall_m])
## Fit train data

# set class_weight for imbalanced classes
class_weight={0:1,1:1,2:1,3:10}


model_glove.fit(X_train, y_train_cat, validation_data=(X_val, y_val_cat), epochs = 30,class_weight=class_weight)

Train on 534 samples, validate on 178 samples
Epoch 1/30
534/534 [==============================] - 4s 7ms/step - loss: 1.7363 - acc: 0.3446 - f1_m: 0.0694 - precision_m: 0.4168 - recall_m: 0.0393 - val_loss: 1.2083 - val_acc: 0.5730 - val_f1_m: 0.3379 - val_precision_m: 0.5528 - val_recall_m: 0.2472
Epoch 2/30
534/534 [==============================] - 1s 2ms/step - loss: 1.4361 - acc: 0.5712 - f1_m: 0.4008 - precision_m: 0.6729 - recall_m: 0.2903 - val_loss: 1.1201 - val_acc: 0.6236 - val_f1_m: 0.3388 - val_precision_m: 0.6301 - val_recall_m: 0.2360
Epoch 3/30
534/534 [==============================] - 1s 2ms/step - loss: 1.2514 - acc: 0.6423 - f1_m: 0.4493 - precision_m: 0.7698 - recall_m: 0.3240 - val_loss: 0.9977 - val_acc: 0.6685 - val_f1_m: 0.5583 - val_precision_m: 0.7461 - val_recall_m: 0.4494
Epoch 4/30
534/534 [==============================] - 1s 2ms/step - loss: 1.1339 - acc: 0.6985 - f1_m: 0.6106 - precision_m: 0.7862 - recall_m: 0.5019 - val_loss: 0.9156 - val_acc: 0.691

In [60]:
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report

y_prob = model_glove.predict(X_test)
y_pred = y_prob.argmax(axis=-1)

confusion_matrix(y_test, y_pred)
print(classification_report(y_test, y_pred))

# can use macro avg as evaluation metric

              precision    recall  f1-score   support

           0       0.80      0.84      0.82        98
           1       0.75      0.72      0.73        53
           2       0.70      0.70      0.70        23
           3       0.00      0.00      0.00         4

   micro avg       0.76      0.76      0.76       178
   macro avg       0.56      0.56      0.56       178
weighted avg       0.75      0.76      0.76       178



## LSTM+glove

In [364]:

X_train, X_val_test, y_train, y_val_test = train_test_split(stroke_train, stroke_label,
                                                    stratify=stroke_label, 
                                                    test_size=0.4,
                                                    random_state=0
                                                   )

X_val,X_test,y_val,y_test = train_test_split(X_val_test, y_val_test,
                                                    stratify=y_val_test, 
                                                    test_size=0.5,
                                                    random_state=0
                                                   )
y_train_cat= np_utils.to_categorical(y_train)
y_test_cat= np_utils.to_categorical(y_test)
y_val_cat= np_utils.to_categorical(y_val)


In [365]:
X_train.shape,X_test.shape,y_train_cat.shape,y_test_cat.shape,

((534, 100), (178, 100), (534, 4), (178, 4))

In [260]:

model_glove = Sequential()
model_glove.add(Embedding(vocabulary_size, 300, input_length=100, weights=[embedding_matrix], trainable=False))
model_glove.add(Dropout(0.2))
# model_glove.add(Conv1D(64, 5, activation='relu'))
# model_glove.add(MaxPooling1D(pool_size=4))
model_glove.add(LSTM(100))
model_glove.add(Dropout(0.2))
model_glove.add(Dense(4, activation='softmax'))
# model_glove.add(Dense(4, activation='sigmoid'))
model_glove.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['acc',f1_m,precision_m, recall_m])
## Fit train data

# set class_weight for imbalanced classes
class_weight={0:1,1:1,2:10,3:20}


model_glove.fit(X_train, y_train_cat, validation_data=(X_val, y_val_cat), epochs = 15,class_weight=class_weight)

Train on 534 samples, validate on 178 samples
Epoch 1/15
 32/534 [>.............................] - ETA: 57s - loss: 3.5664 - acc: 0.0938 - f1_m: 0.0000e+00 - precision_m: 0.0000e+00 - recall_m: 0.0000e+00

KeyboardInterrupt: 

In [259]:
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report

y_prob = model_glove.predict(X_test)
y_pred = y_prob.argmax(axis=-1)

confusion_matrix(y_test, y_pred)
print(classification_report(y_test, y_pred))

# can use macro avg as evaluation metric

              precision    recall  f1-score   support

           0       0.81      0.73      0.77        98
           1       0.89      0.45      0.60        53
           2       0.31      0.83      0.45        23
           3       0.00      0.00      0.00         4

   micro avg       0.65      0.65      0.65       178
   macro avg       0.50      0.50      0.45       178
weighted avg       0.75      0.65      0.66       178



D:\softwares\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1143: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)
D:\softwares\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1143: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)
D:\softwares\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1143: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)


## CNN+glove

In [53]:
from keras.layers import Embedding

word_index = tokenizer.word_index
EMBEDDING_DIM=300
MAX_SEQUENCE_LENGTH=100
embedding_layer = Embedding(len(word_index) + 1,
                            EMBEDDING_DIM,
                            weights=[embedding_matrix],
                            input_length=MAX_SEQUENCE_LENGTH,
                            trainable=False)

In [54]:
from keras.layers import Input
from keras.layers.merge import concatenate
from keras.models import Sequential, Model
from keras.layers import Dense, Embedding, Activation, merge, Input, Lambda, Reshape
from keras.layers import Convolution1D, Flatten, Dropout, MaxPool1D, GlobalAveragePooling1D

# 模型结构：词嵌入-卷积池化*3-拼接-全连接-dropout-全连接
main_input = Input(shape=(100,), dtype='float64')
# 词嵌入（使用预训练的词向量）
embedder = Embedding(len(word_index) + 1, 300, input_length = 100, weights = [embedding_matrix], trainable = False)
embed = embedder(main_input)
# 词窗大小分别为3,4,5
cnn1 = Convolution1D(128, 3, padding='same', strides = 2, activation='relu')(embed)
cnn1 = MaxPool1D(pool_size=5)(cnn1)
cnn2 = Convolution1D(128, 4, padding='same', strides = 2, activation='relu')(embed)
cnn2 = MaxPool1D(pool_size=5)(cnn2)
cnn3 = Convolution1D(128, 5, padding='same', strides = 2, activation='relu')(embed)
cnn3 = MaxPool1D(pool_size=5)(cnn3)
# 合并三个模型的输出向量
# cnn = concatenate([cnn1,cnn2], axis=-1)
cnn = concatenate([cnn1,cnn2,cnn3], axis=-1)
flat = Flatten()(cnn)
drop = Dropout(0.2)(flat)
main_output = Dense(4, activation='softmax')(drop)
model_cnn = Model(inputs = main_input, outputs = main_output)

model_cnn.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['acc',f1_m,precision_m, recall_m])


In [55]:
class_weight={0:1,1:1,2:1,3:10}

model_cnn.fit(X_train, y_train_cat, validation_data=(X_val, y_val_cat), epochs = 30,class_weight=class_weight)

Train on 534 samples, validate on 178 samples
Epoch 1/30
534/534 [==============================] - 3s 5ms/step - loss: 1.6495 - acc: 0.5787 - f1_m: 0.4043 - precision_m: 0.5526 - recall_m: 0.3521 - val_loss: 1.1841 - val_acc: 0.4663 - val_f1_m: 0.2398 - val_precision_m: 0.7551 - val_recall_m: 0.1461
Epoch 2/30
534/534 [==============================] - 1s 1ms/step - loss: 0.9370 - acc: 0.7022 - f1_m: 0.6271 - precision_m: 0.8577 - recall_m: 0.5206 - val_loss: 0.8430 - val_acc: 0.7079 - val_f1_m: 0.6925 - val_precision_m: 0.7483 - val_recall_m: 0.6461
Epoch 3/30
534/534 [==============================] - 1s 1ms/step - loss: 0.6009 - acc: 0.8502 - f1_m: 0.8431 - precision_m: 0.9072 - recall_m: 0.7884 - val_loss: 0.8031 - val_acc: 0.7247 - val_f1_m: 0.7210 - val_precision_m: 0.7770 - val_recall_m: 0.6742
Epoch 4/30
534/534 [==============================] - 1s 1ms/step - loss: 0.3992 - acc: 0.9045 - f1_m: 0.9093 - precision_m: 0.9568 - recall_m: 0.8670 - val_loss: 0.7505 - val_acc: 0.752

In [56]:
# old result
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report

y_prob = model_cnn.predict(X_test)
y_pred = y_prob.argmax(axis=-1)


# confusion_matrix(y_test[int(len(X_test)/2):], y_pred)
print(classification_report(y_test, y_pred))
# can use macro avg as evaluation metric

              precision    recall  f1-score   support

           0       0.71      0.95      0.81        98
           1       0.82      0.62      0.71        53
           2       0.86      0.26      0.40        23
           3       0.00      0.00      0.00         4

   micro avg       0.74      0.74      0.74       178
   macro avg       0.60      0.46      0.48       178
weighted avg       0.75      0.74      0.71       178



D:\softwares\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1143: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)
D:\softwares\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1143: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)
D:\softwares\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1143: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)


## char-embedding

In [123]:
char_map_vec={}

with open('../data/char-embeddings.txt','r') as f:
    for line in f:
        char_map_vec[line[0]]=np.fromstring(line[1:],sep=' ')
print('totally',len(char_map_vec),'chars.')
print("char vector shape:",char_map_vec['a'].shape)

totally 91 chars.
char vector shape: (300,)


In [124]:
max_l=0
l=0

for s in df['tweet'].tolist():
    if(len([c for c in s if c in char_map_vec])>max_l):
        max_l=len([c for c in s if c in char_map_vec])
    
print('max number of chars in twitter:',max_l)


max number of chars in twitter: 402


In [125]:
# in df_map_char_emb, key is the row index of original df file, value is the char embedding of each tweet

df_map_char_emb={}
for i,s in enumerate(df['tweet'].tolist()):
    each_tweet_char_emb=np.zeros((402,300))
    for j,c in enumerate([c for c in s if c in char_map_vec]):
        each_tweet_char_emb[j]=char_map_vec[c]
    df_map_char_emb[i]=each_tweet_char_emb

In [126]:
stroke_train=data[df_to_evaluate.index]
stroke_label=df_to_evaluate.label.tolist()

char_train=[df_map_char_emb[i] for i in df_to_evaluate.index]

In [127]:
print(stroke_train.shape)
char_train=np.array(char_train)
print(char_train.shape)

(890, 100)
(890, 402, 300)


In [128]:

from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
from sklearn.model_selection import train_test_split

X_train, X_val_test, y_train, y_val_test = train_test_split(stroke_train, stroke_label,
                                                    stratify=stroke_label, 
                                                    test_size=0.4,
                                                    random_state=0
                                                   )

X_val,X_test,y_val,y_test = train_test_split(X_val_test, y_val_test,
                                                    stratify=y_val_test, 
                                                    test_size=0.5,
                                                    random_state=0
                                                   )
# print(y_test)
y_train_cat= np_utils.to_categorical(y_train)
y_test_cat= np_utils.to_categorical(y_test)
y_val_cat= np_utils.to_categorical(y_val)


X_char_train, X_char_val_test, y_train, y_val_test = train_test_split(char_train, stroke_label,
                                                    stratify=stroke_label, 
                                                    test_size=0.4,
                                                    random_state=0
                                                   )

X_char_val,X_char_test,y_val,y_test = train_test_split(X_char_val_test, y_val_test,
                                                    stratify=y_val_test, 
                                                    test_size=0.5,
                                                    random_state=0
                                                   )

print(X_char_train.shape,X_char_val.shape,X_char_test.shape,X_train.shape)

(534, 402, 300) (178, 402, 300) (178, 402, 300) (534, 100)


### cnn-char model building

refer to [paper](https://www.aclweb.org/anthology/W19-5025.pdf)  
l1 = char + convo layers  
l2 = word emb + convo layers  
concat l1, l2 to form input 

In [25]:
from keras.layers import Embedding

word_index = tokenizer.word_index
EMBEDDING_DIM=300
MAX_SEQUENCE_LENGTH=100
embedding_layer = Embedding(len(word_index) + 1,
                            EMBEDDING_DIM,
                            weights=[embedding_matrix],
                            input_length=MAX_SEQUENCE_LENGTH,
                            trainable=False)

In [31]:
from keras.layers import Input
from keras.layers.merge import concatenate
from keras.models import Sequential, Model
from keras.layers import Dense, Embedding, Activation, merge, Input, Lambda, Reshape
from keras.layers import Convolution1D, Flatten, Dropout, MaxPool1D, GlobalAveragePooling1D
from keras.layers import Dense, Activation
from keras import backend as K

# part 1 of model: the cnn of the word embeddings
# 模型结构：词嵌入-卷积池化*3-拼接-全连接-dropout-全连接
main_input = Input(shape=(100,), dtype='float64')
# 词嵌入（使用预训练的词向量）
embedder = Embedding(len(word_index) + 1, 300, input_length = 100, weights = [embedding_matrix], trainable = False)
embed = embedder(main_input)
# 词窗大小分别为3,4,5
cnn1 = Convolution1D(128, 3, padding='same', strides = 2, activation='relu')(embed)
cnn1 = MaxPool1D(pool_size=5)(cnn1)
cnn2 = Convolution1D(128, 4, padding='same', strides = 2, activation='relu')(embed)
cnn2 = MaxPool1D(pool_size=5)(cnn2)
cnn3 = Convolution1D(128, 5, padding='same', strides = 2, activation='relu')(embed)
cnn3 = MaxPool1D(pool_size=5)(cnn3)
# 合并三个模型的输出向量

# part 2 of model: the cnn of the char embeddings
char_input=Input(shape=(402,300),dtype='float32')
char_cnn1 = Convolution1D(128, 3, padding='same', strides = 2, activation='relu')(char_input)
char_cnn1 = MaxPool1D(pool_size=5)(char_cnn1)
char_cnn2 = Convolution1D(128, 4, padding='same', strides = 2, activation='relu')(char_input)
char_cnn2 = MaxPool1D(pool_size=5)(char_cnn2)
# char_cnn3 = Convolution1D(128, 5, padding='same', strides = 2, activation='relu')(char_input)
# char_cnn3 = MaxPool1D(pool_size=5)(char_cnn3)


# part 3: integration of the model
cnn = concatenate([cnn1,cnn2,cnn3], axis=-1)
# char_cnn = concatenate([char_cnn1,char_cnn2,char_cnn3], axis=-1)
char_cnn = concatenate([char_cnn1,char_cnn2], axis=-1)
# char_cnn = char_cnn1
flat_1 = Flatten()(cnn)
flat_2 = Flatten()(char_cnn)
flat = concatenate([flat_1,flat_2], axis=-1)
drop = Dropout(0.2)(flat)


main_output = Dense(4, activation='softmax')(drop)
model_cnn = Model(inputs = [main_input,char_input], outputs = main_output)

model_cnn.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['acc',f1_m,precision_m, recall_m])
# model_cnn.summary()

In [303]:
class_weight={0:1,1:1,2:10,3:20}

model_cnn.fit([X_train,X_char_train], y_train_cat, validation_data=([X_val,X_char_val], y_val_cat), epochs = 15,class_weight=class_weight)

Train on 534 samples, validate on 178 samples
Epoch 1/15
534/534 [==============================] - 7s 14ms/step - loss: 3.4294 - acc: 0.1816 - f1_m: 0.1186 - precision_m: 0.1920 - recall_m: 0.0880 - val_loss: 1.5979 - val_acc: 0.1348 - val_f1_m: 0.1203 - val_precision_m: 0.1630 - val_recall_m: 0.0955
Epoch 2/15
534/534 [==============================] - 3s 6ms/step - loss: 2.0047 - acc: 0.4813 - f1_m: 0.3556 - precision_m: 0.6058 - recall_m: 0.2678 - val_loss: 1.6449 - val_acc: 0.1461 - val_f1_m: 0.1575 - val_precision_m: 0.2047 - val_recall_m: 0.1292
Epoch 3/15
534/534 [==============================] - 3s 6ms/step - loss: 1.3122 - acc: 0.6929 - f1_m: 0.6100 - precision_m: 0.8464 - recall_m: 0.4869 - val_loss: 1.1147 - val_acc: 0.5112 - val_f1_m: 0.4014 - val_precision_m: 0.5954 - val_recall_m: 0.3034
Epoch 4/15
534/534 [==============================] - 3s 6ms/step - loss: 0.7756 - acc: 0.8783 - f1_m: 0.8511 - precision_m: 0.9538 - recall_m: 0.7715 - val_loss: 0.9852 - val_acc: 0.65

In [304]:
# old result
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report

y_prob = model_cnn.predict([X_test,X_char_test])
y_pred = y_prob.argmax(axis=-1)


# confusion_matrix(y_test[int(len(X_test)/2):], y_pred)
print(classification_report(y_test, y_pred))
# can use macro avg as evaluation metric

              precision    recall  f1-score   support

           0       0.73      0.93      0.82        98
           1       0.87      0.64      0.74        53
           2       0.57      0.35      0.43        23
           3       0.00      0.00      0.00         4

   micro avg       0.75      0.75      0.75       178
   macro avg       0.54      0.48      0.50       178
weighted avg       0.73      0.75      0.73       178



D:\softwares\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1143: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)
D:\softwares\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1143: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)
D:\softwares\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1143: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)


### lstm (integrate cnn-char) model

In [129]:
from keras.layers import Embedding

word_index = tokenizer.word_index
EMBEDDING_DIM=300
MAX_SEQUENCE_LENGTH=100
embedding_layer = Embedding(len(word_index) + 1,
                            EMBEDDING_DIM,
                            weights=[embedding_matrix],
                            input_length=MAX_SEQUENCE_LENGTH,
                            trainable=False)

In [137]:
# from keras.utils import plot_model
from keras.models import Model
from keras.layers import Input
from keras.layers import Dense
from keras.layers.recurrent import LSTM

# part1- word embeddings
word_input = Input(shape=(100,))
emb=Embedding(vocabulary_size, 300, input_length=100, weights=[embedding_matrix], trainable=False)(word_input)
hidden1 = LSTM(100)(emb)
dropout=Dropout(0.2)(hidden1)


# part2- character embeddings
char_input=Input(shape=(402,300),dtype='float32')
char_cnn1 = Convolution1D(128, 3, padding='same', strides = 2, activation='relu')(char_input)
char_cnn1 = MaxPool1D(pool_size=5)(char_cnn1)
char_cnn2 = Convolution1D(128, 4, padding='same', strides = 2, activation='relu')(char_input)
char_cnn2 = MaxPool1D(pool_size=5)(char_cnn2)
# char_cnn3 = Convolution1D(128, 5, padding='same', strides = 2, activation='relu')(char_input)
# char_cnn3 = MaxPool1D(pool_size=5)(char_cnn3)


#part3- concat all

char_cnn = concatenate([char_cnn1,char_cnn2], axis=-1)
flat_2 = Flatten()(char_cnn)
flat = concatenate([dropout,flat_2], axis=-1)
drop = Dropout(0.2)(flat)

output = Dense(4, activation='softmax')(flat)
char_lstm = Model(inputs=[word_input,char_input], outputs=output)
char_lstm.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['acc',f1_m,precision_m, recall_m])

# summarize layers
# print(char_lstm.summary())
# plot graph
class_weight={0:1,1:1,2:1,3:10}


char_lstm.fit([X_train,X_char_train], y_train_cat, validation_data=([X_val,X_char_val], y_val_cat), epochs = 15,class_weight=class_weight)


Train on 534 samples, validate on 178 samples
Epoch 1/15
534/534 [==============================] - 11s 20ms/step - loss: 1.7236 - acc: 0.4363 - f1_m: 0.1841 - precision_m: 0.5213 - recall_m: 0.1386 - val_loss: 1.2835 - val_acc: 0.5730 - val_f1_m: 0.4823 - val_precision_m: 0.6458 - val_recall_m: 0.3876
Epoch 2/15
534/534 [==============================] - 7s 12ms/step - loss: 1.2440 - acc: 0.6648 - f1_m: 0.6176 - precision_m: 0.8002 - recall_m: 0.5075 - val_loss: 1.0198 - val_acc: 0.6517 - val_f1_m: 0.6111 - val_precision_m: 0.7079 - val_recall_m: 0.5393
Epoch 3/15
534/534 [==============================] - 7s 13ms/step - loss: 0.8825 - acc: 0.7622 - f1_m: 0.7357 - precision_m: 0.8662 - recall_m: 0.6442 - val_loss: 0.8866 - val_acc: 0.7247 - val_f1_m: 0.6810 - val_precision_m: 0.7973 - val_recall_m: 0.5955
Epoch 4/15
534/534 [==============================] - 7s 12ms/step - loss: 0.5758 - acc: 0.8539 - f1_m: 0.8388 - precision_m: 0.9135 - recall_m: 0.7772 - val_loss: 0.8185 - val_acc: 

In [138]:
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report

y_prob = char_lstm.predict([X_test,X_char_test])
y_pred = y_prob.argmax(axis=-1)

confusion_matrix(y_test, y_pred)
print(classification_report(y_test, y_pred))

# can use macro avg as evaluation metric

              precision    recall  f1-score   support

           0       0.77      0.90      0.83        98
           1       0.77      0.70      0.73        53
           2       0.86      0.52      0.65        23
           3       0.00      0.00      0.00         4

   micro avg       0.77      0.77      0.77       178
   macro avg       0.60      0.53      0.55       178
weighted avg       0.76      0.77      0.76       178



### lstm (integrate rnn-char) model

# result table

## old result of binary version

### f1 score of positive class 
(the class with fewer samples)



|   type  | df_alzheimer | df_cancer | df_depression | df_heart_attack |df_parkinson | df_stroke  |
| -------------| ------------- |------------- |------------- |------------- |------------- |------------- |
| MLP  |  0.29  | 0.23  | 0.48  | 0.17  | 0.10  |0.14   |
| LSTM  |  0.61  | 0.39  | 0.52  | 0.48  | 0.53  |0.42  |
| CNN  |  0.42  | 0.32  | 0.48 | 0.27  | 0.14  |0.24  |

### macro average of f1 score
why choose macro average : [csdn](https://www.cnblogs.com/yuuken/p/8822496.html)



|   type  |  df_alzheimer | df_cancer | df_depression | df_heart_attack |df_parkinson |df_stroke  |
| -------------| ------------- |------------- |------------- |------------- |------------- |------------- |
| MLP  |  0.58  | 0.54  | 0.55  | 0.54  | 0.50  |0.50   |
| LSTM  | 0.77  | 0.62  | 0.62  | 0.71  | 0.74 | 0.67  |
| CNN  | 0.66  | 0.60 | 0.59  | 0.60 | 0.54 |0.58  | 

## new result of 4 classes
macro average of f1



|   type  |  df_alzheimer | df_cancer | df_depression | df_heart_attack |df_parkinson |df_stroke  |
| -------------| ------------- |------------- |------------- |------------- |------------- |------------- |
| MLP  |  0.38  |  0.40  | 0.35  | 0.60  | 0.69  |0.56   |
| LSTM  | 0.37  |  0.48  | 0.41  | 0.58  | 0.75 | 0.58  |
| Char-LSTM| 0.37    |   0.49   |   0.40   |  0.50    |  0.73    |   0.57   | 
| CNN  | 0.38  |  0.41 | 0.42  | 0.46 | 0.62 |0.47  | 
| Char-CNN| 0.45|  0.45|  0.40|   0.46|   0.66| 0.50|

# some notes in training:

- Since it's a highly imbalanced dataset in all diseases, we change the loss function by assigning weights to the different classes. For classes with very few samples, we set a relatively high weigh (5-10 times compared to the classes with many samples) 
    - Please refer to [csdn blog](https://blog.csdn.net/notHeadache/article/details/82183503) "带权重的交叉熵loss" 那一节。
    - or refer to [medium](https://towardsdatascience.com/practical-tips-for-class-imbalance-in-binary-classification-6ee29bcdb8a7)-part 4. Class weighted / cost sensitive learning


-  In the original experiments, *self-mention* and *other-mention* labels are taken as **positive class**; and *awareness* and *non-health* labels are taken as **negative class**. The same in this notebook.


- train , val ,test split:  6 : 2 : 2


# conclusion

- char-cnn gives some improvement compared to cnn.
- char-lstm(integrate char-cnn module) worsen the result a little bit.